src/providers/aws.rs

```diff
@@ -0,0 +1,858 @@
+    }
+}
+
+impl Eq for MachineSetup {}
+
+impl std::hash::Hash for MachineSetup {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.instance_type.hash(state);
+        self.ami.hash(state);
+    }
+}
+
+impl Default for MachineSetup {
+    fn default() -> Self {
+        MachineSetup {
+            region: Region::UsEast1,
+            instance_type: "t3.small".into(),
+            ami: UbuntuAmi::from(Region::UsEast1).into(),
+            setup: None,
+        }
+    }
+}
+
+impl MachineSetup {
+    /// Set up the machine in a specific EC2
+    /// [`Region`](http://rusoto.github.io/rusoto/rusoto_core/region/enum.Region.html).
+    ///
+    /// The default region is us-east-1. [Available regions are listed
+    /// here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-available-regions)
+    /// AMIs are region-specific. This will overwrite the ami field to
+    /// the Ubuntu 18.04 LTS AMI in the selected region.
+    pub fn region(mut self, region: Region) -> Self {
+        self.region = region.clone();
+        self.ami = UbuntuAmi::from(region).into();
+        self
+    }
+
+    /// The given AWS EC2 instance type will be used. Note that only [EC2 Defined Duration Spot
+    /// Instance types](https://aws.amazon.com/ec2/spot/pricing/) are allowed.
+    pub fn instance_type(mut self, typ: impl ToString) -> Self {
+        self.instance_type = typ.to_string();
+        self
+    }
+
+    /// The new instance will start out in the state dictated by the Amazon Machine Image specified
+    /// in `ami`. Default is Ubuntu 18.04 LTS.
+    pub fn ami(mut self, ami: impl ToString) -> Self {
+        self.ami = ami.to_string();
+        self
+    }
+
+    /// The `setup` argument is called once for every spawned instances of this type with a handle
+    /// to the target machine. Use [`Machine::ssh`](struct.Machine.html#structfield.ssh) to issue
+    /// commands on the host in question.
+    pub fn setup(
+        mut self,
+        setup: impl Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync + 'static,
+    ) -> Self {
+        self.setup = Some(Arc::new(setup));
+        self
+    }
+}
+
+/// Launch AWS EC2 spot instances.
+///
+/// This implementation uses [rusoto](https://crates.io/crates/rusoto_core) to connect to AWS.
+/// Therefore, you must call [`with_credentials`](AWSLauncher::with_credentials) to authenticate
+/// with AWS.
+/// Each individual region is handled by `AWSRegion`.
+pub struct AWSLauncher<P: ProvideAwsCredentials> {
+    credential_provider: Option<Box<dyn Fn() -> Result<P, Error>>>,
+    max_instance_duration: Option<std::time::Duration>,
+    regions: HashMap<<MachineSetup as super::MachineSetup>::Region, AWSRegion>,
+}
+
+impl<P: ProvideAwsCredentials> Default for AWSLauncher<P> {
+    fn default() -> Self {
+        AWSLauncher {
+            credential_provider: None,
+            max_instance_duration: None,
```

Comment:
This `None` is a little misleading if we (as the docs below indicate) actually default it to 6 hours?

Reply:
there is an `unwrap_or_else(|| 360)` below

Reply:
I guess I'm wondering why have it be an `Option` at all?

Reply:
ah, fair enough.